docs/proposals/custom-metrics-api.md

```diff
@@ -0,0 +1,189 @@
+master metrics API is generally useful to multiple cluster components).
+
+The HPA can currently refer to all pods selected by a label selector, and
+is planned to be able to support referring to metrics describing
+a namespace and other non-pod objects within the namespace.
+
+API Paths
+---------
+
+The root API path will look like `/apis/general-metrics/v1alpha1`.  For
+brevity, this will be left off below.
+
+- `/nodes/{node-name}/metrics/{metric-name...}`: retrieve the given metric
+  on the given node
+
+- `/namespaces/{namespace-name}/metrics/{metric-name...}`: retrieve the
+  given metric on the given namespace
+
+- `/namespaces/{namespace-name}/object-metrics/{object-type}/{metric-name...}`:
+  retrieve the given metric for all objects of the given type in the given
+  namespace.
+
+- `/namespaces/{namespace-name}/object-metrics/{object-type}/{metric-name...}?labelSelector=foo`:
+  retrieve the given metric for all objects of the given type matching the
+  given label selector in the given namespace.
+
+- `/namespaces/{namespace-name}/object-metrics/{object-type}/{metric-name...}?names=foo,bar`:
+  retrieve the given metric for the objects of the given type with the
+  given name in the given namespace
+
+For example, to retrieve the custom metric "hits-per-second" for all pods
+matching "app=frontend` in the namespaces "webapp", the request might look
+like:
+
+`/apis/general-metrics/v1alpha1/namespaces/webapp/object-metrics/pods/hits-per-second?labelSelector=app%3Dfrontend`.
+
+API Objects
+-----------
+
+The request URLs listed above will return objects of the `Metrics`
+type, described below:
+
+```go
+
+// a list of values for a given metric for some set of objects
+type Metrics struct {
+    unversioned.TypeMeta `json:",inline"`
+    unversioned.ListMeta `json:"metadata,omitempty"`
+
+    // the name of the metric
+    MetricName string `json:"metricName"`
+
+    // the value of the metric across the described objects
+    MetricValues []MetricValue `json:"metricValues"`
+}
+
+// a metric value for some object
+type GenericMetrics struct {
+    // the type and object metadata describes the object in question
+    unversioned.TypeMeta `json:",inline"`
+    v1.ObjectMeta        `json:"metadata,omitempty"`
+
+    // indicates the end of the time window containing these metrics (i.e.
+    // these metrics come from some time in [Timestamp-Window, Timestamp])
+    Timestamp unversioned.Time `json:"timestamp"`
+
+    // indicates the duration of the time window containing these metrics
+    Window    unversioned.Duration `json:"window"`
+
+    // the value of the metric for this
+    Value float64
+}
+```
+
+For instance, the example request above would yield the following object:
+
+```json
+{
+    "kind": "Metrics",
+    "apiVersion": "general-metrics/v1alpha1",
+    "metricName": "hits-per-second",
+    "metricValues": [
+        {
+            "kind": "Pod",
+            "metadata": { "name": "server1", "namespace": "webapp" },
```

Comment:
It doesn't currently include it in the master metrics API -- we get to choose which bits an pieces we actually populate, although it isn't expressed at the type level.  It might be advantageous to express the relationship using a objectreference instead, but I was going for symmetry with the master/resource metrics API.  I'm not opposed to changing this, but I have a preference for keeping them relatively similar in shape.